common/src/ipc/transmit.rs

```diff
@@ -288,6 +305,131 @@ impl<T> IpcSocket<T> {
+                code: Code::ReqKill,
+                shm: None,
+            },
+            Request::Clear(clear) => {
+                let mut mmap = Mmap::create(clear.size());
+                clear.serialize(&mut Cursor::new(mmap.slice_mut()));
+                Self {
+                    code: Code::ReqClear,
+                    shm: Some(mmap),
+                }
+            }
+            Request::Img(image) => {
+                let mut mmap = Mmap::create(image.size());
+                image.serialize(&mut Cursor::new(mmap.slice_mut()));
+                Self {
+                    code: Code::ReqImg,
+                    shm: Some(mmap),
+                }
+            }
+        }
+    }
+}
+
+impl<'a> From<&'a IpcMessage> for Request<'a> {
+    fn from(value: &'a IpcMessage) -> Self {
+        match value.code {
+            Code::ReqPing => Self::Ping,
+            Code::ReqQuery => Self::Query,
+            Code::ReqKill => Self::Kill,
+            Code::ReqClear => {
+                let mmap = value.shm.as_ref().expect("clear request must contain data");
+                let clear = ClearRequest::deserialize(&mut Cursor::new(mmap.slice()));
+                Self::Clear(clear)
+            }
+            Code::ReqImg => {
+                let mmap = value.shm.as_ref().expect("image request must contain data");
+                let image = ImageRequest::deserialize(&mut Cursor::new(mmap.slice()));
+                Self::Img(image)
+            }
+            _ => unreachable!("`Request` builder reached invalid state"),
+        }
+    }
+}
+
+impl From<Response> for IpcMessage {
+    fn from(value: Response) -> Self {
+        match value {
+            Response::Ok => Self {
+                code: Code::ResOk,
+                shm: None,
+            },
+            Response::Ping(false) => Self {
+                code: Code::ResAwait,
+                shm: None,
+            },
+            Response::Ping(true) => Self {
+                code: Code::ResConfigured,
+                shm: None,
+            },
+            Response::Info(info) => {
+                let mut mmap = Mmap::create(info.size());
+                info.serialize(&mut Cursor::new(mmap.slice_mut()));
+                Self {
+                    code: Code::ReqImg,
+                    shm: Some(mmap),
+                }
+            }
+        }
+    }
+}
+
+impl From<IpcMessage> for Response {
+    fn from(value: IpcMessage) -> Self {
+        match value.code {
+            Code::ResOk => Self::Ok,
+            Code::ResAwait => Self::Ping(false),
+            Code::ResConfigured => Self::Ping(true),
+            Code::ResInfo => {
+                let mmap = value.shm.as_ref().expect("info request must contain data");
+                let info = Box::<[Info]>::deserialize(&mut Cursor::new(mmap.slice()));
+                Self::Info(info)
+            }
+            _ => unreachable!("`Response` builder reached invalid state"),
```

Comment:
Why is this state invalid? imo I feel like it should at least be mentioned in a comment, if not the message itself, why those are invalid states (and maybe the `unreachable!` could explicitly say what state it's in right now for debugging purposes). If these invalid states could be prevented the type system somehow, though, that would obvs be best.

Reply:
The message consists of code and optional payload. When I read via socket and shove it in `IpcMessage`, I then let this message be interpreted by calling code. If you get a message and expect it to be a `Response`, getting `Request` breaks this expectation. I didn't find elegant way to do this, but at least I will try to improve the diagnostics.

Reply:
If you can at least build a better error message, I think that would greatly improve debugging and figuring out what went wrong when people report issues.